api/src/main/java/org/pmiops/workbench/billing/BillingProjectBufferService.java

```diff
@@ -0,0 +1,62 @@
+package org.pmiops.workbench.billing;
+
+import static org.pmiops.workbench.db.model.BillingProjectBufferEntry.Status.CREATING;
+
+import java.sql.Timestamp;
+import java.time.Clock;
+import java.util.UUID;
+import javax.inject.Provider;
+import org.pmiops.workbench.config.WorkbenchConfig;
+import org.pmiops.workbench.db.dao.BillingProjectBufferEntryDao;
+import org.pmiops.workbench.db.model.BillingProjectBufferEntry;
+import org.pmiops.workbench.firecloud.FireCloudService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+@Service
+public class BillingProjectBufferService {
+
+  private final BillingProjectBufferEntryDao billingProjectBufferEntryDao;
+  private final Clock clock;
+  private final FireCloudService fireCloudService;
+  private final Provider<WorkbenchConfig> workbenchConfigProvider;
+
+  @Autowired
+  public BillingProjectBufferService(BillingProjectBufferEntryDao billingProjectBufferEntryDao,
+      Clock clock,
+      FireCloudService fireCloudService,
+      Provider<WorkbenchConfig> workbenchConfigProvider) {
+    this.billingProjectBufferEntryDao = billingProjectBufferEntryDao;
+    this.clock = clock;
+    this.workbenchConfigProvider = workbenchConfigProvider;
+    this.fireCloudService = fireCloudService;
+  }
+
+  public void bufferBillingProject() {
+    if (getCurrentBufferSize() >= getBufferCapacity()) {
+      return;
+    }
+
+    final String projectName = createBillingProjectName();
+
+    fireCloudService.createAllOfUsBillingProject(projectName);
+    BillingProjectBufferEntry entry = new BillingProjectBufferEntry();
+    entry.setProjectName(projectName);
+    entry.setCreationTime(new Timestamp(clock.instant().toEpochMilli()));
+    entry.setStatus(CREATING);
+
+    billingProjectBufferEntryDao.save(entry);
+  }
+
+  private String createBillingProjectName() {
+    return workbenchConfigProvider.get().firecloud.billingProjectPrefix + "-" + UUID.randomUUID().toString();
```

Comment:
Did we want to make the UUID shorter? i.e. substring this?

Reply:
i believe we decided it was fine to keep the full UUID since its not shown to the user at any point.

Reply:
I might have misunderstood our tech talk conclusion, but I thought we'd move from

https://all-of-us-workbench-test.appspot.com/workspaces/aou-rw-test-324/my-workspace-name

to something like

https://all-of-us-workbench-test.appspot.com/workspaces/aou-rw-test-12345/

e.g. get rid of the workspace name in the URL path, since it's no longer required once project==workspace.

This maintains the nice property that the URL is globally unique, since no two GCP projects can have the same project ID.

Reply:
oh, right.

I was thinking we could still decouple the billing project id === workspace name and allow the user to create a customizable workspace id. Given that the billing project id has to be unique given the current architecture constraints with the buffer.

All that said, i think we decided it was expose the full guid in the url as long as its the last part of the URL. Which it will be in the workspace === billing project model.

Reply:
For the workspace URL itself it will be the last token, but for e.g. notebooks URLs it will still be stuck in the middle. A URL from the test RW today is:

https://all-of-us-workbench-test.appspot.com/workspaces/aou-rw-test-324/asdf/notebooks/Using%2520Cohorts%2520and%2520Concepts.ipynb?playgroundMode=false

The project ID also shows up in the cloud console, and may eventually be used to derive a workspace bucket (which is currently "fc-{GUID}", but could be changed once we have 1-to-1 mapping from project to workspace).

So overall there's still probably some aesthetic pressure to keep the project ID ~short.

The nice thing is, the chance of collision is small and Google will handle rejecting an attempted duplicate project creation. So there's probably no architecture change necessary beyond your already-planned failure tolerance.

(Quick historical note: Firecloud used to solve this for test projects by sticking a few random words at the end, e.g. "fiab-sshah-known-sawfly". It's cute, but random UUID chars are probably safer.)

Reply:
OK. I'm cool with sizing it down in case we end up using it in the future. Thoughts on how big?

Reply:
32 bits of entropy seems sufficient to me indefinitely. Could probably get away with ~24 bits but I don't know if it's worth the optimization (and we might need to increase it later). Supporting up to 100K workspaces comfortably is probably a reasonable enough upper bound to consider for now.